### PR TITLE
deps: update go version to 1.16, bump kubebuilder

### LINKED ISSUE #4828: operator-sdk supports Go version 1.16
#### What did you do?

brew install operator-sdk && operator-sdk init --domain example.com --repo github.com/example/memcached-operator

#### What did you expect to see?

a working operator-sdk

#### What did you see instead? Under which circumstances?

Error: failed to initialize project: unable to run pre-scaffold tasks of "base.go.kubebuilder.io/v3": go version 'go1.16.3' is incompatible because 'requires 1.13 <= version < 1.16'.

#### Environment

**Operator type:**

/language go 

`$ operator-sdk version`

operator-sdk version: "v1.6.2", commit: "b131ca8ec77c96b9898470eba9560c30af0f23f3", kubernetes version: "v1.19.4", go version: "go1.16.3", GOOS: "linux", GOARCH: "amd64

`$ go version` (if language is Go)

go version go1.16.3 linux/amd64

### PR DESCRIPTION
**Pull request by sujil02**
Updating operator-SDK from go1.15 to go 1.16 
Closes #4828

Signed-off-by: sujil02 <sujil.shah@gmail.com>

**Description of the change:**

**Motivation for the change:**

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

**Comment by estroz** [comment]
@sujil02 looks like you need to add a `go mod tidy` to the testdata generator.

File: hack/generate/samples/internal/go/v2/memcached_with_webhooks.go [hunk 1]
```diff
@@ -100,6 +101,10 @@ func (mh *Memcached) Run() {
 	mh.uncommentDefaultKustomization()
 	mh.uncommentManifestsKustomization()
 
+	cmd := exec.Command(mh.ctx.Dir, "go mod tidy")
+	_, err = mh.ctx.Run(cmd)
```
Comment (estroz) [inline hack/generate/samples/internal/go/v2/memcached_with_webhooks.go:105]:
```suggestion
	cmd := exec.Command("go", "mod", "tidy")
	cmd.Dir = mh.ctx.Dir
	_, err = mh.ctx.Run(cmd)
```

**Comment by estroz** [comment]
Looks like ubi:8.3 does not support go1.16 yet. You'll have to download and install it from the golang DL page with `curl`: https://golang.org/dl/

File: hack/generate/samples/internal/go/v2/memcached_with_webhooks.go (hunk 1 above)
Comment (sujil02) [reply hack/generate/samples/internal/go/v2/memcached_with_webhooks.go]: Done

File: hack/generate/samples/internal/go/v3/memcached_with_webhooks.go [hunk 2]
```diff
@@ -100,6 +101,11 @@ func (mh *Memcached) Run() {
 	mh.uncommentDefaultKustomization()
 	mh.uncommentManifestsKustomization()
 
+	cmd := exec.Command("go", "mod", "tidy")
+	cmd.Dir = mh.ctx.Dir
+	_, err = mh.ctx.Run(cmd)
+	pkg.CheckError("Running go mod tidy", err)
+
```
Comment (camilamacedo86) [inline hack/generate/samples/internal/go/v3/memcached_with_webhooks.go:108]: should not be done in KB ? 
Should we upgrade SDK before upgrade KB?

Comment (estroz) [reply hack/generate/samples/internal/go/v3/memcached_with_webhooks.go:108]: Kubebuilder has been upgraded to go 1.16.

File: images/operator-sdk/Dockerfile [hunk 3]
```diff
@@ -19,6 +19,10 @@ RUN GOOS=linux GOARCH=$TARGETARCH make build/operator-sdk
 # Final image.
 FROM registry.access.redhat.com/ubi8/ubi-minimal:8.3
 
+ENV GO_VERSION 1.16.4
+
+RUN curl -O https://storage.googleapis.com/golang/go$GO_VERSION.linux-386.tar.gz
+
 RUN microdnf install -y golang make which
```
Comment (estroz) [inline images/operator-sdk/Dockerfile:26]: Gotta unpack the thing, also I don't think that's the recommended DL link on the go website.
```suggestion
ARG TARGETARCH

RUN microdnf install -y make which tar gzip

ENV GO_VERSION 1.16.4
RUN curl -sSLo /tmp/go.tar.gz https://golang.org/dl/go${GO_VERSION}.linux-${TARGETARCH}.tar.gz \
	&& rm -rf /usr/local/go \
	&& tar -C /usr/local -xzf /tmp/go.tar.gz \
	&& ln -sf /usr/local/go/bin/* /usr/local/bin/ \
	&& rm -f /tmp/go.tar.gz \
	&& go version
```

Comment (sujil02) [reply images/operator-sdk/Dockerfile]: Done. Thank you.

**Pull request merged**